test/hotspot/jtreg/serviceability/jvmti/events/Exception/exception01/libexception01.cpp

```diff
@@ -0,0 +1,317 @@
+    return Agent_Initialize(jvm, options, reserved);
+}
+JNIEXPORT jint JNI_OnLoad_exception01(JavaVM *jvm, char *options, void *reserved) {
+    return JNI_VERSION_1_8;
+}
+#endif
+jint Agent_Initialize(JavaVM *jvm, char *options, void *reserved) {
+  jvmtiError err;
+  jint res;
+  jvmtiCapabilities caps;
+
+  res = jvm->GetEnv((void **) &jvmti_env, JVMTI_VERSION_1_1);
+  if (res != JNI_OK || jvmti_env == NULL) {
+    LOG("Wrong result of a valid call to GetEnv!\n");
+    return JNI_ERR;
+  }
+
+  memset(&caps, 0, sizeof(jvmtiCapabilities));
+  caps.can_generate_exception_events = 1;
+  caps.can_support_virtual_threads = 1;
+
+  err = jvmti_env->AddCapabilities(&caps);
+  if (err != JVMTI_ERROR_NONE) {
+    LOG("(AddCapabilities) unexpected error: %s (%d)\n",
+           TranslateError(err), err);
+    return JNI_ERR;
+  }
+
+  err = jvmti_env->GetCapabilities(&caps);
+  if (err != JVMTI_ERROR_NONE) {
+    LOG("(GetCapabilities) unexpected error: %s (%d)\n",
+           TranslateError(err), err);
+    return JNI_ERR;
+  }
+
+  if (caps.can_generate_exception_events) {
+    callbacks.Exception = &Exception;
+    err = jvmti_env->SetEventCallbacks(&callbacks, sizeof(callbacks));
+    if (err != JVMTI_ERROR_NONE) {
+      LOG("(SetEventCallbacks) unexpected error: %s (%d)\n",
+             TranslateError(err), err);
+      return JNI_ERR;
+    }
+  } else {
+    LOG("Warning: Exception event is not implemented\n");
+  }
+
+  return JNI_OK;
+}
+
+JNIEXPORT jint JNICALL
+Java_exception01_check(JNIEnv *jni, jclass cls) {
+  jvmtiError err;
+  jthread thread;
+  jclass clz;
+  jmethodID mid;
+
+  if (jvmti_env == NULL) {
+    LOG("JVMTI client was not properly loaded!\n");
+    return STATUS_FAILED;
+  }
+
+  clz = jni->FindClass("exception01c");
+  if (clz == NULL) {
+    LOG("Cannot find exception01c class!\n");
+    return STATUS_FAILED;
+  }
+  clz = jni->FindClass("exception01b");
+  if (clz == NULL) {
+    LOG("Cannot find exception01b class!\n");
+    return STATUS_FAILED;
+  }
+  clz = jni->FindClass("exception01a");
+  if (clz == NULL) {
+    LOG("Cannot find exception01a class!\n");
+    return STATUS_FAILED;
+  }
+  mid = jni->GetStaticMethodID(clz, "run", "()V");
+  if (mid == NULL) {
+    LOG("Cannot find method run!\n");
+    return STATUS_FAILED;
+  }
+
+  err = jvmti_env->GetCurrentThread(&thread);
+  if (err != JVMTI_ERROR_NONE) {
+    LOG("Failed to get current thread: %s (%d)\n", TranslateError(err), err);
+    result = STATUS_FAILED;
+    return STATUS_FAILED;
+  }
+
+  err = jvmti_env->SetEventNotificationMode(JVMTI_ENABLE,
+                                            JVMTI_EVENT_EXCEPTION, thread);
+  if (err == JVMTI_ERROR_NONE) {
+    eventsExpected = sizeof(exs)/sizeof(exceptionInfo);
+  } else {
+    LOG("Failed to enable JVMTI_EVENT_EXCEPTION: %s (%d)\n",
+           TranslateError(err), err);
+    result = STATUS_FAILED;
+  }
+
+  eventsCount = 0;
```

Comment:
Counters are not protected with locks.
I'm not sure it is a real problem here but would be better to double-check.

Reply:
The variable eventsCount seems to be used in the sinlge thread only. I looked on the several tests with same variable and similar logic.